File: scripts/process-csv.mjs
import { readFileSync, writeFileSync } from "fs";
import { join, dirname } from "path";
import { fileURLToPath } from "url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const csvPath = join(__dirname, "raw-data.csv");
const outPath = join(__dirname, "..", "lib", "real-data.json");

// ---- CSV Parser that handles quoted multiline fields ----
function parseCSV(text) {
  const rows = [];
  let i = 0;
  const len = text.length;

  // Read header
  let headerEnd = text.indexOf("\n");
  const headers = text.slice(0, headerEnd).trim().split(",");

  i = headerEnd + 1;

  while (i < len) {
    const row = [];
    for (let col = 0; col < headers.length; col++) {
      if (i >= len) { row.push(""); continue; }
      if (text[i] === '"') {
        // Quoted field
        i++; // skip opening quote
        let val = "";
        while (i < len) {
          if (text[i] === '"') {
            if (i + 1 < len && text[i + 1] === '"') {
              val += '"'; i += 2;
            } else {
              i++; // skip closing quote
              break;
            }
          } else {
            val += text[i]; i++;
          }
        }
        row.push(val);
        if (i < len && text[i] === ",") i++; // skip comma
        else if (i < len && (text[i] === "\n" || text[i] === "\r")) {
          if (text[i] === "\r" && i + 1 < len && text[i + 1] === "\n") i += 2;
          else i++;
        }
      } else {
        // Unquoted field
        let val = "";
        while (i < len && text[i] !== "," && text[i] !== "\n" && text[i] !== "\r") {
          val += text[i]; i++;
        }
        row.push(val);
        if (i < len && text[i] === ",") i++;
        else if (i < len && (text[i] === "\n" || text[i] === "\r")) {
          if (text[i] === "\r" && i + 1 < len && text[i + 1] === "\n") i += 2;
          else i++;
        }
      }
    }
    if (row.length >= 5 && row[0]) rows.push(row);
  }

  return { headers, rows };
}

console.log("Reading CSV...");
const raw = readFileSync(csvPath, "utf-8");
console.log(`CSV size: ${(raw.length / 1024 / 1024).toFixed(1)} MB`);

const { headers, rows } = parseCSV(raw);
console.log(`Parsed ${rows.length} rows, ${headers.length} columns`);
console.log("Headers:", headers);

// Column indices
const COL = {};
headers.forEach((h, i) => (COL[h.trim()] = i));

const TS = COL["Timestamp"];
const SRC_IP = COL["Source IP Address"];
const DST_IP = COL["Destination IP Address"];
const SRC_PORT = COL["Source Port"];
const DST_PORT = COL["Destination Port"];
const PROTOCOL = COL["Protocol"];
const PKT_LEN = COL["Packet Length"];
const PKT_TYPE = COL["Packet Type"];
const TRAFFIC = COL["Traffic Type"];
const MALWARE = COL["Malware Indicators"];
const ANOMALY = COL["Anomaly Scores"];
const ALERTS = COL["Alerts/Warnings"];
const ATTACK_TYPE = COL["Attack Type"];
const ATTACK_SIG = COL["Attack Signature"];
const ACTION = COL["Action Taken"];
const SEVERITY = COL["Severity Level"];
const USER_INFO = COL["User Information"];
const DEVICE = COL["Device Information"];
const SEGMENT = COL["Network Segment"];
const GEO = COL["Geo-location Data"];
const LOG_SRC = COL["Log Source"];

// ---- Aggregations ----
const totalEvents = rows.length;

// Count maps
const protocolCount = {};
const attackTypeCount = {};
const severityCount = {};
const actionCount = {};
const segmentCount = {};
const trafficTypeCount = {};
const packetTypeCount = {};
const logSourceCount = {};
const attackSigCount = {};

// Time series by month
const monthlyEvents = {};
const monthlySeverity = {};

// Anomaly score stats
let anomalySum = 0;
let anomalyCount = 0;
let anomalyMax = 0;
const anomalyBuckets = { "0-20": 0, "20-40": 0, "40-60": 0, "60-80": 0, "80-100": 0 };

// Alerts
let alertTriggered = 0;
let iocDetected = 0;

// Packet length stats
let pktLenSum = 0;
let pktLenCount = 0;

// Unique IPs
const uniqueSrcIPs = new Set();
const uniqueDstIPs = new Set();

// Top source IPs
const srcIPCount = {};

// Geo locations
const geoCount = {};

// Sample rows for table (grab 500 evenly distributed)
const sampleIndices = new Set();
const step = Math.max(1, Math.floor(rows.length / 500));
for (let s = 0; s < rows.length; s += step) sampleIndices.add(s);

const sampleRows = [];

for (let r = 0; r < rows.length; r++) {
  const row = rows[r];
  const timestamp = row[TS];
  const protocol = row[PROTOCOL] || "Unknown";
  const attackType = row[ATTACK_TYPE] || "None";
  const severity = row[SEVERITY] || "Unknown";
  const action = row[ACTION] || "Unknown";
  const segment = row[SEGMENT] || "Unknown";
  const trafficType = row[TRAFFIC] || "Unknown";
  const packetType = row[PKT_TYPE] || "Unknown";
  const logSource = row[LOG_SRC] || "Unknown";
  const attackSig = row[ATTACK_SIG] || "Unknown";
  const anomalyScore = parseFloat(row[ANOMALY]) || 0;
  const pktLen = parseInt(row[PKT_LEN]) || 0;
  const alert = (row[ALERTS] || "").trim();
  const malware = (row[MALWARE] || "").trim();
  const srcIP = row[SRC_IP] || "";
  const dstIP = row[DST_IP] || "";
  const geo = (row[GEO] || "").trim();

  // Increment counts
  protocolCount[protocol] = (protocolCount[protocol] || 0) + 1;
  attackTypeCount[attackType] = (attackTypeCount[attackType] || 0) + 1;
  severityCount[severity] = (severityCount[severity] || 0) + 1;
  actionCount[action] = (actionCount[action] || 0) + 1;
  segmentCount[segment] = (segmentCount[segment] || 0) + 1;
  trafficTypeCount[trafficType] = (trafficTypeCount[trafficType] || 0) + 1;
  packetTypeCount[packetType] = (packetTypeCount[packetType] || 0) + 1;
  logSourceCount[logSource] = (logSourceCount[logSource] || 0) + 1;
  attackSigCount[attackSig] = (attackSigCount[attackSig] || 0) + 1;

  // Anomaly
  if (anomalyScore > 0) {
    anomalySum += anomalyScore;
    anomalyCount++;
    if (anomalyScore > anomalyMax) anomalyMax = anomalyScore;
    if (anomalyScore < 20) anomalyBuckets["0-20"]++;
    else if (anomalyScore < 40) anomalyBuckets["20-40"]++;
    else if (anomalyScore < 60) anomalyBuckets["40-60"]++;
    else if (anomalyScore < 80) anomalyBuckets["60-80"]++;
    else anomalyBuckets["80-100"]++;
  }

  // Alerts
  if (alert === "Alert Triggered") alertTriggered++;
  if (malware === "IoC Detected") iocDetected++;

  // Packet length
  if (pktLen > 0) { pktLenSum += pktLen; pktLenCount++; }

  // IPs
  if (srcIP) { uniqueSrcIPs.add(srcIP); srcIPCount[srcIP] = (srcIPCount[srcIP] || 0) + 1; }
  if (dstIP) uniqueDstIPs.add(dstIP);

  // Geo
  if (geo) {
    const state = geo.split(",").pop()?.trim() || geo;
    geoCount[state] = (geoCount[state] || 0) + 1;
  }

  // Monthly
  if (timestamp) {
    const month = timestamp.slice(0, 7); // "2023-05"
    monthlyEvents[month] = (monthlyEvents[month] || 0) + 1;
    if (!monthlySeverity[month]) monthlySeverity[month] = { Low: 0, Medium: 0, High: 0 };
    if (severity === "Low" || severity === "Medium" || severity === "High") {
      monthlySeverity[month][severity]++;
    }
  }

  // Sample
  if (sampleIndices.has(r)) {
    sampleRows.push({
      id: `EVT-${String(r + 1).padStart(5, "0")}`,
      date: timestamp,
      srcIp: srcIP,
      dstIp: dstIP,
      srcPort: parseInt(row[SRC_PORT]) || 0,
      dstPort: parseInt(row[DST_PORT]) || 0,
      protocol,
      packetLength: pktLen,
      packetType,
      trafficType,
      attackType,
      attackSignature: attackSig,
      severity,
      action,
      anomalyScore,
      segment,
      geo,
      logSource,
      malwareIndicator: malware,
      alert,
      user: row[USER_INFO] || "",
      device: (row[DEVICE] || "").slice(0, 60),
    });
  }
}

// Sort monthly
const timeSeries = Object.entries(monthlyEvents)
  .sort(([a], [b]) => a.localeCompare(b))
  .map(([month, count]) => ({
    month,
    total: count,
    low: monthlySeverity[month]?.Low || 0,
    medium: monthlySeverity[month]?.Medium || 0,
    high: monthlySeverity[month]?.High || 0,
  }));

// Top source IPs
const topSrcIPs = Object.entries(srcIPCount)
  .sort((a, b) => b[1] - a[1])
  .slice(0, 10)
  .map(([ip, count]) => ({ ip, count }));

// Top geos
const topGeos = Object.entries(geoCount)
  .sort((a, b) => b[1] - a[1])
  .slice(0, 10)
  .map(([location, count]) => ({ location, count }));

// Build output
const output = {
  summary: {
    totalEvents,
    uniqueSrcIPs: uniqueSrcIPs.size,
    uniqueDstIPs: uniqueDstIPs.size,
    alertsTriggered: alertTriggered,
    iocDetected,
    avgAnomalyScore: anomalyCount > 0 ? Math.round((anomalySum / anomalyCount) * 100) / 100 : 0,
    maxAnomalyScore: Math.round(anomalyMax * 100) / 100,
    avgPacketLength: pktLenCount > 0 ? Math.round(pktLenSum / pktLenCount) : 0,
  },
  distributions: {
    protocol: protocolCount,
    attackType: attackTypeCount,
    severity: severityCount,
    action: actionCount,
    segment: segmentCount,
    trafficType: trafficTypeCount,
    packetType: packetTypeCount,
    logSource: logSourceCount,
    attackSignature: attackSigCount,
    anomalyBuckets,
  },
  timeSeries,
  topSrcIPs,
  topGeos,
  sampleRows,
};

writeFileSync(outPath, JSON.stringify(output, null, 2), "utf-8");
console.log(`Output written to ${outPath}`);
console.log("Summary:", JSON.stringify(output.summary, null, 2));
console.log(`Time series months: ${timeSeries.length}`);
console.log(`Sample rows: ${sampleRows.length}`);
console.log("Done!");

File: scripts/process-csv.py
import csv
import json
import sys
from collections import Counter, defaultdict
from io import StringIO

csv_path = "/vercel/share/v0-project/scripts/raw-data.csv"
out_path = "/vercel/share/v0-project/lib/real-data.json"

print(f"Reading CSV from {csv_path}...")
with open(csv_path, "r", encoding="utf-8") as f:
    raw = f.read()

print(f"CSV size: {len(raw) / 1024 / 1024:.1f} MB")
reader = csv.DictReader(StringIO(raw))
rows_list = list(reader)
total = len(rows_list)
print(f"Parsed {total} rows")

protocol_c = Counter()
attack_c = Counter()
severity_c = Counter()
action_c = Counter()
segment_c = Counter()
traffic_c = Counter()
pkt_type_c = Counter()
log_src_c = Counter()
atk_sig_c = Counter()
monthly = defaultdict(int)
monthly_sev = defaultdict(lambda: {"Low": 0, "Medium": 0, "High": 0})
anomaly_sum = 0.0
anomaly_count = 0
anomaly_max = 0.0
anomaly_buckets = {"0-20": 0, "20-40": 0, "40-60": 0, "60-80": 0, "80-100": 0}
alerts_triggered = 0
ioc_detected = 0
pkt_sum = 0
pkt_count = 0
unique_src = set()
unique_dst = set()
src_ip_c = Counter()
geo_c = Counter()

step = max(1, total // 500)
sample = []

for i, row in enumerate(rows_list):
    ts = row.get("Timestamp", "").strip()
    protocol = row.get("Protocol", "Unknown").strip() or "Unknown"
    attack = row.get("Attack Type", "None").strip() or "None"
    sev = row.get("Severity Level", "Unknown").strip() or "Unknown"
    act = row.get("Action Taken", "Unknown").strip() or "Unknown"
    seg = row.get("Network Segment", "Unknown").strip() or "Unknown"
    traffic = row.get("Traffic Type", "Unknown").strip() or "Unknown"
    pkt_type = row.get("Packet Type", "Unknown").strip() or "Unknown"
    log_src = row.get("Log Source", "Unknown").strip() or "Unknown"
    atk_sig = row.get("Attack Signature", "Unknown").strip() or "Unknown"
    anom_raw = row.get("Anomaly Scores", "0").strip()
    pkt_raw = row.get("Packet Length", "0").strip()
    alert = row.get("Alerts/Warnings", "").strip()
    malware = row.get("Malware Indicators", "").strip()
    src_ip = row.get("Source IP Address", "").strip()
    dst_ip = row.get("Destination IP Address", "").strip()
    geo = row.get("Geo-location Data", "").strip()

    try:
        anomaly = float(anom_raw)
    except:
        anomaly = 0.0
    try:
        pkt_len = int(pkt_raw)
    except:
        pkt_len = 0

    protocol_c[protocol] += 1
    attack_c[attack] += 1
    severity_c[sev] += 1
    action_c[act] += 1
    segment_c[seg] += 1
    traffic_c[traffic] += 1
    pkt_type_c[pkt_type] += 1
    log_src_c[log_src] += 1
    atk_sig_c[atk_sig] += 1

    if anomaly > 0:
        anomaly_sum += anomaly
        anomaly_count += 1
        anomaly_max = max(anomaly_max, anomaly)
        if anomaly < 20: anomaly_buckets["0-20"] += 1
        elif anomaly < 40: anomaly_buckets["20-40"] += 1
        elif anomaly < 60: anomaly_buckets["40-60"] += 1
        elif anomaly < 80: anomaly_buckets["60-80"] += 1
        else: anomaly_buckets["80-100"] += 1

    if alert == "Alert Triggered": alerts_triggered += 1
    if malware == "IoC Detected": ioc_detected += 1
    if pkt_len > 0: pkt_sum += pkt_len; pkt_count += 1
    if src_ip: unique_src.add(src_ip); src_ip_c[src_ip] += 1
    if dst_ip: unique_dst.add(dst_ip)

    if geo:
        parts = geo.split(",")
        state = parts[-1].strip() if len(parts) > 1 else geo
        geo_c[state] += 1

    if ts:
        month = ts[:7]
        monthly[month] += 1
        if sev in ("Low", "Medium", "High"):
            monthly_sev[month][sev] += 1

    if i % step == 0 and len(sample) < 500:
        sample.append({
            "id": f"EVT-{i+1:05d}",
            "date": ts,
            "srcIp": src_ip,
            "dstIp": dst_ip,
            "srcPort": int(row.get("Source Port", "0").strip() or "0"),
            "dstPort": int(row.get("Destination Port", "0").strip() or "0"),
            "protocol": protocol,
            "packetLength": pkt_len,
            "packetType": pkt_type,
            "trafficType": traffic,
            "attackType": attack,
            "attackSignature": atk_sig,
            "severity": sev,
            "action": act,
            "anomalyScore": round(anomaly, 2),
            "segment": seg,
            "geo": geo,
            "logSource": log_src,
            "malwareIndicator": malware,
            "alert": alert,
            "user": row.get("User Information", "").strip(),
            "device": row.get("Device Information", "").strip()[:60],
        })

time_series = sorted(
    [{"month": m, "total": c, "low": monthly_sev[m]["Low"], "medium": monthly_sev[m]["Medium"], "high": monthly_sev[m]["High"]}
     for m, c in monthly.items()],
    key=lambda x: x["month"]
)

top_src = [{"ip": ip, "count": c} for ip, c in src_ip_c.most_common(10)]
top_geos = [{"location": g, "count": c} for g, c in geo_c.most_common(10)]

output = {
    "summary": {
        "totalEvents": total,
        "uniqueSrcIPs": len(unique_src),
        "uniqueDstIPs": len(unique_dst),
        "alertsTriggered": alerts_triggered,
        "iocDetected": ioc_detected,
        "avgAnomalyScore": round(anomaly_sum / anomaly_count, 2) if anomaly_count > 0 else 0,
        "maxAnomalyScore": round(anomaly_max, 2),
        "avgPacketLength": round(pkt_sum / pkt_count) if pkt_count > 0 else 0,
    },
    "distributions": {
        "protocol": dict(protocol_c),
        "attackType": dict(attack_c),
        "severity": dict(severity_c),
        "action": dict(action_c),
        "segment": dict(segment_c),
        "trafficType": dict(traffic_c),
        "packetType": dict(pkt_type_c),
        "logSource": dict(log_src_c),
        "attackSignature": dict(atk_sig_c),
        "anomalyBuckets": anomaly_buckets,
    },
    "timeSeries": time_series,
    "topSrcIPs": top_src,
    "topGeos": top_geos,
    "sampleRows": sample,
}

with open(out_path, "w", encoding="utf-8") as f:
    json.dump(output, f, indent=2, ensure_ascii=False)

print(f"Output written to {out_path}")
print(f"Summary: {json.dumps(output['summary'], indent=2)}")
print(f"Time series months: {len(time_series)}")
print(f"Sample rows: {len(sample)}")
print("Done!")
